File: Data Structures/Arrays/Hashing/LeetCode/1TwoSum/1TwoSum.go
package main

func twoSum(nums []int, target int) []int {
	// Create a map to store numbers and their indices
	lookup := make(map[int]int)

	// Iterate through the array
	for i, num := range nums {
		// Calculate the complement needed to reach the target
		complement := target - num

		// Check if the complement exists in the map
		if index, found := lookup[complement]; found {
			// If found, return the indices of the complement and the current number
			return []int{index, i}
		}

		// Store the current number and its index in the map
		lookup[num] = i
	}

	// Return an empty slice if no solution is found (problem guarantees a solution)
	return []int{}
}

File: Data Structures/Arrays/Hashing/LeetCode/1TwoSum/1TwoSum.cpp
// This solution uses an unordered hashmap 
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> lookup; // Stores value-to-index mapping

        for (int i = 0; i < nums.size(); i++) {
            int complement = target - nums[i]; // Calculate the complement

            // Check if the complement exists in the map
            if (lookup.find(complement) != lookup.end()) {
                // If found, return the indices of the complement and the current number
                return {lookup[complement], i};
            }

            // Store the current number and its index in the map
            lookup[nums[i]] = i;
        }

        // Return an empty vector if no solution is found (problem guarantees one solution)
        return {};
    }
};

// this 
